### ISSUE TITLE
Implementation - for loops and semantics?

### ISSUE THREAD

**Issue by MatAtBread** [tc39/proposal-async-await#29]
Great proposal. You can try it out at https://www.npmjs.com/package/nodent

**Comment by MatAtBread** [comment]
I'm trying to understand the formal semantics of await inside a for-loop, for example:

```
for (var i=0; i<3; i++) {
    console.log(await addOne(i)) ;
    console.log(" next") ;
}

console.log("done") ;
```

Does this output "done 1 next 2 next 3 next" or "1 next 2 next 3 next done"?

The question is really about whether the for loop block is itself considered blocking because it contains blocking statements.

The comparison with the unwound loop would suggest that it is:

```
console.log(await addOne(0)) ;
console.log(" next") ;
console.log(await addOne(1)) ;
console.log(" next") ;
console.log(await addOne(2)) ;
console.log(" next") ;
console.log("done") ;
```

This clearly suggests "done" is output after all the calls have completed. On the other hand, the semantics of blocking just the for loop have the advantage of allow for 'parallel' execution without additional language constructs.

Any input on resolving this would be gratefully received.

**Comment by sebmck** [comment]
@MatAtBread `1 next 2 next 3 next done` is the expected result. `await` is "blocking" regardless of whether it's in a loop or not.

See [await\* and parallelism](https://github.com/lukehoban/ecmascript-asyncawait#await-and-parallelism) for suggested syntax that may be introduced to do promises in parallel. Currently the only way to do parallel awaits is with `Promise.all`:

``` javascript
await Promise.all([
  promise1,
  promise2,
  promise3
]);
```

**Comment by matAtWork** [comment]
Thx. Interesting issue. I'm also trying to get my head aroung what happens in a switch statement when one of the cases contains an await, and one doesn't:

```
switch (x) {
case 1: y = await addOne(x) ; break ;
case 2: y = -1 ; break ;
}
console.log(y);
```

It would appear that the requirement is to hoist all the post-switch statements into a callback that in case 2 is invoked synchronously, rather than asynchronously?

**Comment by domenic** [comment]
"Sync" vs. "async" is not the right POV, because in no cases will execution actually be blocked. But yes, in your switch example, `x = 1` => `console.log(x + 1)`, whereas `x = 2` => `console.log(-1)`.

**Comment by matAtWork** [comment]
OOI, is there a forum where these issues are being discussed/examined?

**Comment by domenic** [comment]
You found it :). But there's not much to examine, as the semantics are all pretty clear from the readme and the desugaring.

**Comment by matAtWork** [comment]
I appreciate there's no blocking. It was more a comment on the possibly of reaching the final console.log(y) directly or via deferred execution depending on the code path, and ultimately a run-time decision (e.g. the value of x)

**Comment by matAtWork** [comment]
I'm working on an alternate implementation with the 'standard' semantics, and came across the for/while/do & switch issues (which probably aren't really 'issues'). @sebmck (above) commented on how there was no language support for parallel async invocations and suggested Promises.all. My implementation doesn't rely on Promises or Generators, just vanilla ES5 & syntax transformation, and has no run-time support, so it would strike me that the language should have a construct for such semantics, even if everyone (except me) implements it under-the-hood via Promises.

In y dev branch, I'm comparing:

```
for (var i=0; i<3; i++) {
    console.log(await addOne(i)) ;
    console.log(" next") ;
}

console.log("done") ;
// "1 next 2 next 3 next done"
```

with

```
async for (var i=0; i<3; i++) {
    console.log(await addOne(i)) ;
    console.log(" next") ;
}

console.log("done") ;
// "done 1 next 2 next 3 next" (or potentially any order)
```

The semantic difference here being that preceding the loop with `async` only makes the body 'blocking', but continues execution after the loop whether or not the async functions have completed.

This provides the language with the semantic capability to implement the various async map/parallel/waterfall (or whatever) without specifying the implementation. I realise it's kind of academic if Promises make it into everything someday, but I'm interested in the language semantics rather than an specific implementation at this point.

FYI, the actually syntax (i.e. the preceding `async`) is not the important bit, if you think it's ugly - it's giving the language the facility to start parallel async tasks I think would of benefit.

**Comment by zenparsing** [comment]
@matAtWork A single execution context cannot have more than one "resume" point, which is what your semantics would imply.

**Comment by MatAtBread** [comment]
Hmmm. I'm not skilled enough to understand exactly what you mean, however it's implementable on ES5 (my transpiler does it). Can you explain it to me I'm a bit more detail? If helps I'm happy to share my current implementation

**Comment by domenic** [comment]
@matAtWork The semantics of this proposal are inherently Promise based, so I am not sure what you are trying to accomplish. Perhaps you want to make a new proposal with its own GitHub repo? This proposal is literally just syntax sugar for creating and reacting to promise-returning functions.

**Comment by zenparsing** [comment]
@MatAtBread what does your transpiler output for the `async for` example?

**Comment by arv** [comment]
One possible interpretation of `async for` would be to run all the iterations of the loop and gather them in an iterable and then Promise.all on it. Something like this:

http://goo.gl/JTZafV

**Comment by matAtWork** [comment]
@domenic I think the proposal is great. I have some questions about whether the implementation reliance on Promises and Generators is necessary or desirable. The original project solved the same issues using a slightly different syntax (co-incidentally) a couple of years back, when ES6 was a twinkle. As a programmer, doing the same mental syntax transformations over and over again made it an obvious case for a programmatic operation. The project has a repo, but that's why I asked if here was a good place to ask questions, as that's how I hope to get a better understanding of the bits I've not yet understood.

**Comment by matAtWork** [comment]
@zenparsing 

Before:

```
async function x(y) {
    await setImmediate ;
    return (y+1) ;
}

async for (var i=0; i<3; i++) {
    console.log("x",await x(i)) ;
}
console.log("x-done") ;
```

After:

```
function x(y) {
    return function($return, $error) {
        try {
            return setImmediate(function($await_setImmediate$1) {
                return $return(y + 1);
            }.bind(this), $error);
        } catch ($except) {
            $error($except)
        }
    }.bind(this);
}

for (var i = 0; i < 3; i++) {
    x(i)(function($await_x$2) {
        console.log("x", $await_x$2);
    }.bind(this), $error);
    continue;
}

console.log("x-done");
```

If you're interested, the 'normal' async/await (outside of a loop) does pretty much what's in the spec, as I currently understand it (which is a work in progress)

**Comment by domenic** [comment]
So unlike the spec, your `async function`s do not always return promises?

**Comment by MatAtBread** [comment]
No, as I said, the transpiler doesn't use any runtime that's not in ES5. I take your point tho, I'm interpreting the spec from the language/keywords/semantics PoV, not the implementation.

**Comment by domenic** [comment]
I think we have very different interpretations of what "semantics" means. Two specs which cause functions with the same syntactic form to return different values at runtime _definitely_ have different semantics. So I do not think you are following any part of the spec except for syntax.

**Comment by zenparsing** [comment]
@arv @MatAtBread  right, you'd have to essentially spawn off an execution contexts for each loop iteration.  But that means that it's no longer obvious where the "current" execution point of the async function is.

I think it's just better to explicitly create closures, if that's what you want.

**Comment by MatAtBread** [comment]
Possibly... I'm not trying to be contentious for the sake of it. My understanding of semantics is the set of definitions that make up a "meaning". In the same way that "I am Matt" and "Je suis Matt" have the same semantics, but differing grammar and syntax (although not by much).

I read the proposal as a "language" proposal defining new keywords and what programmatic constructs they implied, rather than a specific implementation, although I completely see the value in making a proposal with an implementation attached as it makes it easier to understand what the author is describing.

**Comment by arv** [comment]
@zenparsing I agree. The thing that I find most weird about the strawman is that the execution flow is not longer straight forward. For example, the following case would not work as one might expect.

``` js
var i = 0;
async while (i < 10) {
  await i;
  i++;
}
```

This would never terminate.

I therefore agree with you and we need to wait for a while so that we can see how closures are used to  solve them same use case.

**Comment by MatAtBread** [comment]
@zenparsing Sorry to sound dumb. Can you tell me what you mean by execution contexts? There's only one JS activation context within the loop, but it is called-back more than once (once for each iteration).

This kind of goes back to the previous comment about whether the proposal is for language extensions or a specific implementation. I was certainly pushing at the latter when suggesting that the keywords could be used to introduce a structured parallelism into the language, which I don't see as related to the presence of a closure or not

**Comment by domenic** [comment]
@MatAtBread

In programming language design, semantics means something very different than vague "meaning". You seem to use the word "implementation" in place of computer-science "semantics," but again that is incorrect usage: an implementation is the specific code in V8, SpiderMonkey, etc. that implements the specific semantics. (In this case, it is the code that parses the source code, creates a function object in memory that returns a promise, etc.)

Spec proposals have to specify semantics, including behavior (e.g. "what is the return type of this function?"), otherwise different browsers will implement different semantics and authors will not be able to rely on behavior cross-browser. So it would be incorrect to read this proposal as a syntax-only proposal.

Anyway, now that I understand you're not actually interested in this proposal, but instead in your own, I'll back off :). Just hoped it would be helpful for you to understand how these terms are used.

**Comment by MatAtBread** [comment]
@domenic In all honesty, I've not read an LRM since K&R many years ago. I'm interested in challenging and improving the proposal, it's true, but I'm not proposing my own.

If the proposal didn't specify an implementation, it wouldn't break code anymore than if it did - the 'await' keyword accepts as an argument a reference to an object returned by a function marked as 'async'. 

This is no more constraining that saying the identifier preceding parentheses must refer to a 'function', without specifying whether it's implemented as a pointer, selector, handle, or indeed a string (as the early JS interpreters used).

Actually, it's the additional constraint that this object must be a Promise that introduces fragility - a future implantation based on some OS or platform feature would now be impossible.

**Comment by zenparsing** [comment]
@MatAtBread i don't think anyone interested in programming languages sounds "dumb".  : )

An execution context is a kind of stack frame, but it may be suspended (as in the case of generators).  It carries whatever information is necessary to resume at the point of last suspension.

In your strawman, you have multiple resume points (one for each loop iteration).  That's what gives it "parallelism".  This example will make that clear:

``` js
async for (let i of [1, 2, 3]) {
    await opA();
    await opB();
}
```

Each loop iteration may be suspended at either `opA` or `opB`, so we would need to store that state for each loop iteration.

It's doable if you essentially create a new activation for each loop iteration (e.g. by calling a closure), but that interacts poorly with the design.  The whole point of async and await is to be able to program asynchronous stuff sequentially.

It's much more clear what the intention is if you write it explicitly:

``` js
await Promise.all([1, 2, 3].map(async i => {
    await opA();
    await opB();
}));
```

If that's too wordy, you can create a library to do some of the abstraction for you.

``` js
import { mapJoin } from "async-tools";
await mapJoin(iterable, async x => { /*...*/ });
```

**Comment by MatAtBread** [comment]
@zenparsing Thanks for the informative response. Yes, the transpiler generates a closure for each invocation. I guess what's making me "uncomfortable" about the "spec vs implementation" is that the keywords are specified to rely on a runtime (Promises) which the programmer is meant to extend into common use cases (eg parallelism) rather than taking the opportunity to specify the language constructs which meet those common cases and leaving the compiler to work out the best implementation. 

I'm kind of surprised I've found it so hard to articulate this as it seems to have been seen as some kind of "attack" on the proposal (which in general I think is great), which itself accepts the case for parallelism and suggests a syntax to address that case.

Interestingly, it sounds like some of the parallelism issues are related to some of the (intentional) restrictions on Promises, such as only ever resolving once. One of the great things about JS is it's activation record which is much easier to use than the old C setjmp/longjmp, which seems to be analogous to the generator/yield pattern.

Luckily for me, our ES5 approach works on pretty much everything right now, so we choose how we deliver or services. My next task is to benchmark the various implementations and see how they compare.

I really appreciate your insight - thanks again

**Comment by sebmck** [comment]
@MatAtBread The reason it's been seen as an attack is because you're trying to remove promises from the equation when they're a core component.

**Comment by MatAtBread** [comment]
Ah OK. I didn't read it that way but as a set of "language features". 

You're not wrong about me questioning promises, but it's not an attack! If someone questions the use of a resource in a bit of code and provides some demonstration that it is indeed unnecessary, to some that's a step forward or an optimisation.

If the language extensions are there to support Promises, rather than the other way around, then maybe they're not great extensions? Luckily in our game there's space for lots of solutions.

**Comment by MatAtBread** [comment]
It's been an interesting investigation from my PoV, and I've learned some interesting things. One of which is traceur's output (running on Node 10.32, no 'harmony') runs about twice as fast as our transpiler output, which is clearly significant. The test isn't exhaustive - it's focused on the invocation/response performance (10000 loops round a await-async-setImmediate) as opposed to real life tasks like HTTP, etc, and Nodent's loop handling is not very smart.

Traceur's output is somewhat bigger (and harder to follow IMHO) and requires a runtime, but the loop completes (on my machine) in around 920ms, as compared to 1650-1700ms for the ES5 output from nodent.

For those who took the time to engage and talk me through the proposal and thinking - thanks for your time. 

Original test:

```
function immediately(ms) {
  return new Promise((resolve) => {
    setImmediate(resolve);
  });
}

async function x(y) { 
  await immediately() ;
  return y+1; 
}

async function test(n) {
  var t = Date.now() ;
  var i = 0 ;
  while (i<n) {
      i = await x(i) ;
  }
  console.log(i,Date.now()-t) ;
}

test(100000) ;
```

**Comment by matAtWork** [comment]
@domenic @sebmck @zenparsing 
FYI, I updated my transpiler to (optionally) use Promises as opposed to an abstract type (actually a pair of callbacks). This has the advantage of being able to `await` on a Promise (rather than my abstract type), but the disadvantage of requiring run-time support for Promises and a nastier debugging experience (as async invocations from await step in and out of the Promise library).

Still no generators as it uses AST transformation. I'm guessing that your interest is limited, but I've found it really useful and interesting to try and implement this proposal from a minimal perspective.

Here's an example, just for your viewing pleasure. 

Before:

```
async function inc(y) {
    return y+1 ;
}

async function test() {
    console.log("start...") ;
    console.log("done ",await inc(1)) ;
    console.log("...end") ;
    return ;
}

await test() ;
```

After

```
function inc(y) {
    return new Promise(function($return, $error) {
        try {
            return $return(y + 1);
        } catch ($except) {
            return $error($except);
        }
    });
}

function test() {
    return new Promise(function($return, $error) {
        try {
            console.log("start...");
            return inc(1).then(function($await_inc$1) {
                console.log("done ", $await_inc$1);
                console.log("...end");
                return $return();
            }.bind(this), $error);
        } catch ($except) {
            return $error($except);
        }
    });
}

return test().then(function() {}, $error);
```

**Comment by domenic** [comment]
@matAtWork there's no need for the try/catch inside the promise constructor, that is done automatically by `new Promise`.

**Comment by MatAtBread** [comment]
Thx. I'll optimise that one away. It's only needed for the non-Promise version
